File: meldanya-python3/day1/day1.py
def solve(ns, predicate):
    filtered = []
    i = 0
    for i in range(len(ns)):
        if predicate(i, ns):
            filtered.append(ns[i])
        i += 1
    return sum(filtered)


def part1(i, ns):
    return ns[i] == ns[(i+1) % len(ns)]


def part2(i, ns):
    return ns[i] == ns[(i + int(len(ns)/2)) % len(ns)]


def main(inp):
    ns = [int(i) for i in str(inp)]
    print(solve(ns, part1))
    print(solve(ns, part2))


if __name__ == '__main__':
    main(input())

File: zegl-go/day1/day1.go
package day1

import (
	"strings"
	"strconv"
)

func Part1(input string) int {
	sum := 0

	var nums []int

	for _, nStr := range strings.Split(input, "") {
		if n, err := strconv.Atoi(nStr); err == nil {
			nums = append(nums, n)
		}
	}

	// Add first as last
	nums = append(nums, nums[0])

	for i, n := range nums {
		if i == len(nums)-1 {
			break
		}

		if n == nums[i+1] {
			sum += n
		}
	}

	return sum
}

func Part2(input string) int {
	sum := 0

	var nums []int

	for _, nStr := range strings.Split(input, "") {
		if n, err := strconv.Atoi(nStr); err == nil {
			nums = append(nums, n)
		}
	}

	halfLen := len(nums) / 2

	// Add first half to the end
	nums = append(nums, nums[0:halfLen]...)

	for i, n := range nums {
		if i+halfLen < len(nums) && n == nums[i+halfLen] {
			sum += n
		}
	}

	return sum
}
